app/models/labware_creators/donor_pooling_plate.rb

```diff
@@ -0,0 +1,225 @@
+# frozen_string_literal: true
+
+module LabwareCreators
+  # This labware creator receives barcodes for a configured number of source
+  # plates from the user. It pools samples from the passed wells into a
+  # destination plate. It's used for scRNA Donor Pooling to create 'LRC PBMC
+  # Pools' plates from 'LRC PBMC Defrost PBS' plates.
+  #
+  # The creator imposes restrictions:
+  # - It doesn't allow combining samples from different studies or projects.
+  # - It doesn't allow samples with the same donor_id in the same pool.
+  #
+  # The number of pools is determined by a lookup table based on sample count.
+  # Tag depth index is added to aliquot attributes to avoid tag clashes.
+  class DonorPoolingPlate < Base
+    include LabwareCreators::CustomPage
+    include SupportParent::PlateOnly
+
+    include LabwareCreators::DonorPoolingCalculator
+    include LabwareCreators::DonorPoolingValidator
+
+    # The name of the template that will be used for rendering the barcode
+    # input page.
+    self.page = 'donor_pooling_plate'
+
+    # Add the barcodes attribute to the list of attributes for this class.
+    # It is used by the creation controller to permit the barcodes parameter.
+    self.attributes += [{ barcodes: [] }]
```

Comment:
The base class does this to configure what params to permit from input forms: 
self.attributes = %i[purpose_uuid parent_uuid user_uuid]
It will be used for filtering the form parameters sent like this: params.permit(attributes)

In the subclass, you want to add an additional fields to be allowed, like this: self.attributes += [:barcode, :name, :surname, :filename] . That would mean there is a form field named barcode and it has a single value, etc. Each of them refers to a separate parameter.

Now you want to allow an array of values for a param. You need to specify it as a `key: value` where value is an empty array because that is the syntax: barcodes: [] . You now need to add this mapping into your attributes array in the subclass. You use hash syntax to wrap the mapping:  { barcodes: [] } so you can add it to your attributes array. If you had additional params, for example { filenames: [] } , it would be self.attributes += [ { barcodes: [] }, { filenames: [] } ]

permit call will look like 
params.permit(:purpose_uuid, :parent_uuid, :user_uuid, barcodes: [], filenames: [] )

Reply:
OK got it, thanks for the detailed explanation! 

I'd read it wrong, I was thinking that `+=` acts the same as `<<`, but of course it doesn't. So either of the following would work:
```
attributes += [{ barcodes: [] }]
attributes << { barcodes: [] }
```

Reply:
attributes << { barcodes: [] } would work for a single item. For multiple items, it would need chaining: 

`attributes << { barcodes: [] } << { thing_names: [] } << { other_thing_names: [] }`

However, attributes += [{ barcodes: [] }] reads better and that is why labware creators use this syntax:

`attributes += [ { barcodes: [] }, { thing_names: [] }, { other_thing_names: [] } ]`